examples/datastreams/main.go

```diff
@@ -0,0 +1,143 @@
+package main
+
+import (
+	"flag"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/livekit/protocol/logger"
+	lksdk "github.com/livekit/server-sdk-go/v2"
+)
+
+const (
+	sender   = "sender"
+	receiver = "receiver"
+)
+
+var (
+	host, apiKey, apiSecret string
+)
+
+func init() {
+	flag.StringVar(&host, "host", "", "livekit server host")
+	flag.StringVar(&apiKey, "api-key", "", "livekit api key")
+	flag.StringVar(&apiSecret, "api-secret", "", "livekit api secret")
+}
+
+func main() {
+	logger.InitFromConfig(&logger.Config{Level: "info"}, "ds-demo")
+	lksdk.SetLogger(logger.GetLogger())
+	flag.Parse()
+
+	roomName := "datastreams-demo-" + uuid.New().String()[:8]
+	senderRoom, err := lksdk.ConnectToRoom(host, lksdk.ConnectInfo{
+		APIKey:              apiKey,
+		APISecret:           apiSecret,
+		RoomName:            roomName,
+		ParticipantIdentity: sender,
+	}, nil)
+	if err != nil {
+		logger.Errorw("failed to connect to room", err, "participant", sender)
+		return
+	}
+
+	receiverRoom, err := lksdk.ConnectToRoom(host, lksdk.ConnectInfo{
+		APIKey:              apiKey,
+		APISecret:           apiSecret,
+		RoomName:            roomName,
+		ParticipantIdentity: receiver,
+	}, nil)
+	if err != nil {
+		logger.Errorw("failed to connect to room", err, "participant", receiver)
+		return
+	}
+
+	defer func() {
+		senderRoom.Disconnect()
+		receiverRoom.Disconnect()
+	}()
+
+	receiverRoom.RegisterTextStreamHandler("text-receive-iter", func(reader *lksdk.TextStreamReader, participantIdentity string) {
+		logger.Infow("received text stream", "participant", participantIdentity)
+		// Q. If I iterate over the reader directly without a goroutine, I don't receive any dc messages
+		go func() {
```

Comment:
This does not make for a good public API design, how can I handle this?

Reply:
quoted the wrong thing, but basically, I *need* to iterate over the channel inside a goroutine or I stop receiving messages from the data channel (don't know about other messages, but chunks and trailers don't arrive at all)

Reply:
can any required go routines  be moved to internal implementation? Do not fully understand the flow, but yeah, requiring a goroutine for this is not great.

Reply:
Can't do that in the current implementation but, we talked about changing the read/write APIs to more closely resemble https://pkg.go.dev/io, will try to handle it there.

Reply:
This is handled internally now